packages/grafana-flamegraph/src/FlameGraph/dataTransform.ts

```diff
@@ -107,12 +86,67 @@ export function nestedSetToLevels(
     if (parent) {
       parent.children.push(newItem);
     }
-    parent = newItem;
 
+    parent = newItem;
     levels[currentLevel].push(newItem);
   }
 
-  return [levels, uniqueLabels, collapsedMap];
+  const collapsedMapContainer = new CollapsedMapContainer(options?.collapsingThreshold);
+  if (options?.collapsing) {
+    // We collapse similar items here, where it seems like parent and child are the same thing and so the distinction
+    // isn't that important. We create a map of items that should be collapsed together.
+    collapsedMapContainer.addTree(levels[0][0]);
+  }
+
+  return [levels, uniqueLabels, collapsedMapContainer.getMap()];
+}
+
+export type CollapsedMap = Map<LevelItem, CollapseConfig>;
+export class CollapsedMapContainer {
+  private map = new Map();
+  private threshold = 0.99;
+
+  constructor(threshold?: number) {
+    if (threshold !== undefined) {
+      this.threshold = threshold;
+    }
+  }
+
+  addTree(root: LevelItem) {
```

Comment:
Because we need to know how many children item has we need to wait for the whole tree to be done and than do another pass over it.